### PR TITLE
feat: cluster wide startup hook for tasks

### PR DESCRIPTION
**Pull request by hamidzr**
## TODO
- [x] slurm test
- [x] docs and release notes
- [x] an e2e tests that shows the hook running. via logs? or command
## Description

- (sep from this PR) should we be merging RM TCDs with clusterwide TCDs instead of overriding them?
https://github.com/determined-ai/determined/blob/d5f807da0ecded7d7b4c6afbf37bbfce8d0233f2/master/internal/rm/kubernetesrm/kubernetes_resource_manager.go#L631

## Test Plan

- set startup hook various task container defaults: master config, resource manager
  - check that RM startuphook override cluster wide one
- unit tests

basic example
```
        task_container_defaults:
          startup_hook: |
            echo "TCD startup thing from master config"
            echo "task type is $DET_TASK_TYPE"
```

```
hbp ➜ d/determined-ee git:(hz-ee-startuphook) ✗ det cmd run 'ls $HOME; date; ls /run/determined/'
Launched command (id: 3d1d922b-4edd-4556-b2df-a94806138a4a).
[2024-04-09T21:50:07.432659Z]          || INFO: Scheduling Command (singularly-enormous-ibex) (id: 3d1d922b-4edd-4556-b2df-a94806138a4a.1)
[2024-04-09T21:50:11.340487Z]          || INFO: HPC Job ID: 3
[2024-04-09T21:50:11.342877Z]          || INFO: Command (singularly-enormous-ibex) was assigned to an agent
[2024-04-09T21:50:16.739730Z] [cont=0] || WARNING: [10466] determined: falling back to naive proxy ip resolution (error=None)
[2024-04-09T21:50:17.238141Z] [cont=0] || + test -f /run/determined/dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.238237Z] [cont=0] || + source /run/determined/dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.238266Z] [cont=0] || ++ echo hi from master tcd slurm
[2024-04-09T21:50:17.238279Z] [cont=0] || + set +x
[2024-04-09T21:50:17.238379Z] [cont=0] || hi from master tcd slurm
[2024-04-09T21:50:17.246344Z] [cont=0] || Tue Apr  9 21:50:17 UTC 2024
[2024-04-09T21:50:17.248080Z] [cont=0] || command-entrypoint.sh
[2024-04-09T21:50:17.248151Z] [cont=0] || dynamic-tcd-startup-hook.sh
[2024-04-09T21:50:17.248172Z] [cont=0] || etc
[2024-04-09T21:50:17.248190Z] [cont=0] || info
[2024-04-09T21:50:17.248208Z] [cont=0] || pythonuserbase
[2024-04-09T21:50:17.248223Z] [cont=0] || ship-logs.sh
[2024-04-09T21:50:17.248234Z] [cont=0] || ship_logs.py
[2024-04-09T21:50:17.248260Z] [cont=0] || singularity-entrypoint-wrapper.sh
[2024-04-09T21:50:17.248267Z] [cont=0] || task-setup.sh
[2024-04-09T21:50:17.248273Z] [cont=0] || train
[2024-04-09T21:50:17.248279Z] [cont=0] || workdir
[2024-04-09T21:50:21.314010Z]          || INFO: resources exited successfully with a zero exit code
[2024-04-09T21:50:21.330339Z]          || INFO: Command (singularly-enormous-ibex) was terminated: allocation stopped early after all resources exited: resources exited successfully with a zero exit code
Task log stream ended. To reopen log stream, run: det task logs -f 3d1d922b-4edd-4556-b2df-a94806138a4a
```

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

https://hpe-aiatscale.atlassian.net/browse/RM-105

### CODE REVIEW

File: master/pkg/model/task_container_defaults.go [hunk 1]
```diff
@@ -170,6 +172,7 @@ var mergeCopier = copier.Option{IgnoreEmpty: true, DeepCopy: true}
 func (c TaskContainerDefaultsConfig) Merge(
 	other TaskContainerDefaultsConfig,
 ) (TaskContainerDefaultsConfig, error) {
+	// Q: where is this used other than tests?
```
Comment (hamidzr) [inline master/pkg/model/task_container_defaults.go:175]: https://github.com/determined-ai/determined/blob/d5f807da0ecded7d7b4c6afbf37bbfce8d0233f2/master/pkg/schemas/merge.go#L66 ?

File: master/pkg/tasks/task.go [hunk 2]
```diff
@@ -238,6 +238,18 @@ func (t TaskSpec) EnvVars() map[string]string {
 	return e
 }
 
+func (t *TaskSpec) tcdStartHookPath() string {
+	return filepath.Join(RunDir, StartupHookScript)
+}
+
+func (t *TaskSpec) tCDStartupEntrypoint() []string {
```
Comment (hamidzr) [inline master/pkg/tasks/task.go:245]: how do we snakeCase these when it's an acronym

**Review by NicholasBlaskey — COMMENTED** [review]
one comment that I think should provide a simplification

File: master/static/srv/task-setup.sh [hunk 3]
```diff
@@ -53,3 +53,37 @@ if [ "$HOME" = "/" ]; then
     )" || HOME="$PWD"
     export HOME
 fi
+
+cleaned_args=()
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/static/srv/task-setup.sh:57]: that's much simpler, I think I was just excited to write some bash arg parsing I'll update

[the remark this answers is not in the document]
Comment (hamidzr) [reply master/static/srv/task-setup.sh:57]: is it safe to assume task RunDir is fixed? we have a single constant `RunDir` in our go code for it

File: master/static/srv/notebook-entrypoint.sh [hunk 4]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-source /run/determined/task-setup.sh
+source /run/determined/task-setup.sh "$@"
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/static/srv/notebook-entrypoint.sh:3]: now that the logic is [simpler](https://github.com/determined-ai/determined/pull/9124#discussion_r1557851960) I could take these out of task-setup and put it right with where we call startup-hook.sh

File: master/static/srv/task-setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/static/srv/task-setup.sh]: @BOsterbuhr any comments based on your usage and requirements?

**Comment by BOsterbuhr** [comment]
@hamidzr we need to update `helm/charts/determined/templates/master-config.yaml` as well as `helm/charts/determined/values.yaml` to allow for this to be configured via helm.

**Comment by hamidzr** [comment]
> @hamidzr we need to update `helm/charts/determined/templates/master-config.yaml` as well as `helm/charts/determined/values.yaml` to allow for this to be configured via helm.

thanks for the snippet and the reminder. I started working on that [here](https://github.com/determined-ai/determined/compare/hz-global-startup...hz-startup-hook-helm?expand=1)

File: e2e_tests/tests/cluster/test_logging.py [hunk 5]
```diff
@@ -143,6 +145,37 @@ def do_check_logs() -> None:
...
+def test_tcd_startup_hook_task_master() -> None:
+    for ntsc in ["command", "notebook", "shell"]:
+        _test_task_logs(ntsc, {}, re.compile("^.*hello from master tcd startup hook.*$"))
+
+
+@pytest.mark.e2e_cpu
+@pytest.mark.e2e_cpu_elastic
+@pytest.mark.e2e_cpu_cross_version
+@pytest.mark.e2e_gpu  # Note, e2e_gpu and not gpu_required hits k8s cpu tests.
+@pytest.mark.e2e_slurm
```
Comment (NicholasBlaskey) [inline e2e_tests/tests/cluster/test_logging.py:164]: I don't know if you have already but would check this passes in slurm

File: master/static/srv/command-entrypoint.sh [hunk 6]
```diff
@@ -10,6 +10,10 @@ set -e
 # to register the proxy with the Determined master.
 "$DET_PYTHON_EXECUTABLE" -m determined.exec.prep_container --proxy --download_context_directory
 
+set -x
```
Comment (NicholasBlaskey) [inline master/static/srv/command-entrypoint.sh:13]: do commands not support start up hook?

maybe just make a ticket calling this out? Not sure if that is intentional

File: docs/reference/deploy/master-config-reference.rst [hunk 7]
```diff
@@ -103,6 +103,12 @@ specified here. You can customize environment variables for CUDA (NVIDIA GPU), C
 GPU) tasks differently by specifying a dict with ``cuda`` (``gpu`` prior to Determined 0.17.6),
 ``cpu``, and ``rocm`` keys.
 
+``startup_hook``
+================
+
+An optional inline script that will be executed as part of task set up. This is defined under
```
Comment (NicholasBlaskey) [inline docs/reference/deploy/master-config-reference.rst:109]: I know you had a seperate pr but helm / helm docs?

If you wanna do seperate that is fine but I think I prefer one PR

**Review by NicholasBlaskey — APPROVED** [review]
LGTM, nice job

File: master/static/srv/command-entrypoint.sh (hunk 6 above)
Comment (hamidzr) [reply master/static/srv/command-entrypoint.sh:13]: https://hpe-aiatscale.atlassian.net/browse/RM-159

File: e2e_tests/tests/cluster/test_logging.py (hunk 5 above)
Comment (hamidzr) [reply e2e_tests/tests/cluster/test_logging.py:164]: I manually tested the older version and I'm checking the new one. any good way of adding this to be run on slurm from this repo? would need a separate follow up pr on ee after we land this or setup a different ee pr and make modifications there.

File: docs/reference/deploy/master-config-reference.rst (hunk 7 above)
Comment (hamidzr) [reply docs/reference/deploy/master-config-reference.rst:109]: that'll take me longer to test potentially not until tomorrow. I think it'd make sense to make sure the main changes are tested and landed asap?

**Pull request merged**